### ISSUE TITLE
Don't run `tool` scripts until scene is ready and values are populated

### ISSUE THREAD

**Issue by Shadowblitz16** [godotengine/godot-proposals#1539]
**Describe the project you are working on:**
A spaceship game

**Describe the problem or limitation you are having in your project:**
I keep running into annoying issues with godot where the scene is not ready when tool scripts are ran

**Describe the feature / enhancement and how it helps to overcome the problem or limitation:**
I suggest the following...
- properties are not serialized unless they are auto properties and then only their backed fields are
- tool scripts are not ran until the scene is ready and the fields are populated into the script (in that order)
- when scripts are saved they are paused until populated again

**Describe how your proposal will work, with code, pseudocode, mockups, and/or diagrams:**
basically this solve issues with exported properties throwing errors when they access nodes in tool scripts

**If this enhancement will not be used often, can it be worked around with a few lines of script?:**
idk but its not very beginner friendly if there is.

**Is there a reason why this should be core and not an add-on in the asset library?:**
Its really annoying to have to null check every time you use a node that might not be ready.
I think the would be a good way to keep GDScript easy to use and would make it so much less or a pain to write tool scripts

**Comment by KoBeWi** [comment]
Does #325 solve the same problem?

**Comment by TheDuriel** [comment]
It don't see what this solves. Functionality would be lost here as well.

> when they access nodes in tool scripts

You need to be aware of init order in the editor _exactly_ the same way you need to be in your games.


#325 is an example of a one line fix for the rare instances in which it matters to you. (For refernce, the setter in a Node in Tool mode will be called twice. Once on Init with the default value set in the script, and once right after enter_tree() with the value from the scene file.)

```
func somesetter(value):
    if is_inside_tree():
        #do_tree_stuff
```

**Comment by Shadowblitz16** [comment]
my issues is the nodes are not ready when the tool script is reloaded or ran for the first time
also I'm not sure why the editor needs to set the values more then once. it could be just deseralized after the scene is ready.

the idea would to be to do the following with tool scripts when loaded...
- enters scene tree
- deserializes fields
- calls tree_enter
- does this for the rest of the nodes
- when tree is ready call ready

**Comment by TheDuriel** [comment]
That is actually what it is doing.

**Comment by Shadowblitz16** [comment]
but its triggering setters when its deserializing into fields?

**Comment by TheDuriel** [comment]
Because many, many, Nodes and Resources need to run code when those values are set. Most resources for example will generate the "actual" internal data, the export field is just an input for that.

There is no use saving something that can be safely recreated from an easy to serialize and human editable input.

You can't just skip calling the setter when populating the fields.

**Comment by Shadowblitz16** [comment]
so my big question is how does unity do this without things being null

**Comment by Shadowblitz16** [comment]
sorry misclick

**Comment by jonbonazza** [comment]
Can you provide a concrete example of what you are trying to do? Perhaps even provide a minimal reproducible project?

**Comment by Shadowblitz16** [comment]
@jonbonazza  
I am trying to access nodes in properties setters
I don't have a minimal reproducible project but its s simple as trying to get a node in a exported property in a tool script.

the only solution to this problem is to null check every time I accesses a node.
the problem is that properties are being set before the script is ready

TheDuriel's way of doing `if is_inside_tree(): only works if the current node is inside the tree and having to do `is_inside_tree()` for every node your using increases the amount of code we have to write alot.

**Comment by cgbeutler** [comment]
This looks very similar to https://github.com/godotengine/godot/issues/30460
In that issue, bojidar-bg recommends [this style](https://github.com/godotengine/godot/issues/30460#issuecomment-509697259):
```coffeescript
export var x = 5  setget set_x
func set_x(new_x):
  x = new_x
  if is_inside_tree():
    # something like:
    $sprite.frame = x

func _ready():
  set_x(x)
```

This may also be related to the fact that "ready" stuff is not re-called on tool scripts, so onready variables are not reliable.

**Comment by Shadowblitz16** [comment]
@KoBeWi  onready setters work but I would rather just having onready properties as a whole as well as support for onready export properties

**Comment by Shadowblitz16** [comment]
closing in favor of https://github.com/godotengine/godot-proposals/issues/2613
